Return-Path: <benh@kernel.crashing.org>
X-Original-To: olh@wotan.suse.de
Received: from hermes.suse.de (hermes.suse.de [149.44.160.1])
	by wotan.suse.de (Postfix) with ESMTP id 8773D2A8370
	for <olh@wotan.suse.de>; Sat, 21 May 2005 04:47:42 +0200 (CEST)
Received: by hermes.suse.de (Postfix)
	id 81BC716C156; Sat, 21 May 2005 04:47:42 +0200 (CEST)
Received: from scanhost.suse.de (scanhost.suse.de [149.44.160.36])
	by hermes.suse.de (Postfix) with ESMTP id 7B70F16BF0E
	for <olh@suse.de>; Sat, 21 May 2005 04:47:42 +0200 (CEST)
Received: from hermes.suse.de ([149.44.160.1])
 by scanhost.suse.de (scanhost [149.44.160.36]) (amavisd-new, port 10025)
 with ESMTP id 15873-04 for <olh@suse.de>;
 Sat, 21 May 2005 04:47:40 +0200 (CEST)
Received: from mx1.suse.de (mx1.suse.de [195.135.220.2])
	(using TLSv1 with cipher EDH-RSA-DES-CBC3-SHA (168/168 bits))
	(No client certificate requested)
	by hermes.suse.de (Postfix) with ESMTP id F0E3016C462
	for <olh@suse.de>; Sat, 21 May 2005 04:47:37 +0200 (CEST)
Received: from gate.crashing.org (gate.crashing.org [63.228.1.57])
	(using TLSv1 with cipher DHE-RSA-AES256-SHA (256/256 bits))
	(No client certificate requested)
	by mx1.suse.de (Postfix) with ESMTP id 7335CD4E8
	for <olh@suse.de>; Sat, 21 May 2005 04:47:37 +0200 (CEST)
Received: from gaston (localhost [127.0.0.1])
	by gate.crashing.org (8.12.8/8.12.8) with ESMTP id j4L2eMZn031552;
	Fri, 20 May 2005 21:40:23 -0500
Subject: Re: oops in i2c_keywest_init
From: Benjamin Herrenschmidt <benh@kernel.crashing.org>
To: Owen Stampflee <ostampflee@terrasoftsolutions.com>,
	Olaf Hering <olh@suse.de>
Cc: linuxppc64-dev@ozlabs.org
In-Reply-To: <20050520155048.GA31591@suse.de>
References: <20050520155048.GA31591@suse.de>
Content-Type: text/plain
Date: Sat, 21 May 2005 12:45:46 +1000
Message-Id: <1116643546.5153.160.camel@gaston>
Mime-Version: 1.0
X-Mailer: Evolution 2.2.2 
Content-Transfer-Encoding: 7bit
X-Virus-Scanned: by amavisd-new at scanhost.suse.de
X-Spam-Status: No, hits=-1.0 tagged_above=-20.0 required=5.0 tests=BAYES_50,
 MY_LINUX
X-Spam-Level: 

Ok, it looks like the device-tree is bogus on these and lacks interrupt
informations. That plus the i2c-keywest driver doesn't properly test for
the presence of an interrupt in the device node and crashes if there is
none. Can you test this patch ? I haven't even tried building so it may
need some fixup. It tries to "fixup" the device-tree at boot, and also
adds some guard to i2c-keywest. Let me know if it fixes booting and if
i2c works propertly.

Signed-off-by: Olaf Hering <olh@suse.de>
Index: linux-2.6.12-rc4-olh/arch/ppc64/kernel/prom_init.c
===================================================================
--- linux-2.6.12-rc4-olh.orig/arch/ppc64/kernel/prom_init.c
+++ linux-2.6.12-rc4-olh/arch/ppc64/kernel/prom_init.c
@@ -1750,7 +1750,42 @@ static void __init flatten_device_tree(v
 	prom_printf("Device tree struct  0x%x -> 0x%x\n",
 		    RELOC(dt_struct_start), RELOC(dt_struct_end));
 
- }
+}
+
+
+static void __init fixup_device_tree(void)
+{
+	unsigned long offset = reloc_offset();
+	phandle u3, i2c, mpic;
+	u32 u3_rev;
+	u32 interrupts[2];
+	u32 parent;
+
+	/* Some G5s have a missing interrupt definition, fix it up here */
+	u3 = call_prom("finddevice", 1, 1, ADDR("/u3@0,f8000000"));
+	if ((long)u3 <= 0)
+		return;
+	i2c = call_prom("finddevice", 1, 1, ADDR("/u3@0,f8000000/i2c@f8001000"));
+	if ((long)i2c <= 0)
+		return;
+	mpic = call_prom("finddevice", 1, 1, ADDR("/u3@0,f8000000/mpic@f8040000"));
+	if ((long)mpic <= 0)
+		return;
+
+	if (prom_getprop(u3, "device-rev", &u3_rev, sizeof(u3_rev)) <= 0)
+		return;
+	if (u3_rev != 0x35)
+		return;
+	if (prom_getproplen(i2c, "interrupts") >= 0)
+		return;
+	/* interrupt on this revision of u3 is number 0 and level */
+	interrupts[0] = 0;
+	interrupts[1] = 1;
+	prom_setprop(i2c, "interrupts", &interrupts, sizeof(interrupts));
+	parent = (u32)mpic;
+	prom_setprop(i2c, "interrupt-parent", &parent, sizeof(parent));
+}
+
 
 static void __init prom_find_boot_cpu(void)
 {
@@ -1920,6 +1955,11 @@ unsigned long __init prom_init(unsigned 
 	}
 
 	/*
+	 * Fixup any known bugs in the device-tree
+	 */
+	fixup_device_tree();
+
+	/*
 	 * Now finally create the flattened device-tree
 	 */
        	prom_printf("copying OF device tree ...\n");
Index: linux-2.6.12-rc4-olh/drivers/i2c/busses/i2c-keywest.c
===================================================================
--- linux-2.6.12-rc4-olh.orig/drivers/i2c/busses/i2c-keywest.c
+++ linux-2.6.12-rc4-olh/drivers/i2c/busses/i2c-keywest.c
@@ -516,6 +516,11 @@ create_iface(struct device_node *np, str
 	u32 *psteps, *prate;
 	int rc;
 
+	if (np->n_intrs < 1 || np->n_addrs < 1) {
+		printk(KERN_ERR "%s: Missing interrupt or address !\n",
+		       np->full_name);
+		return -ENODEV;
+	}
 	if (pmac_low_i2c_lock(np))
 		return -ENODEV;
 
